### PR TITLE
[Detect before initialize] POAE7-288: Detect configuration before initialize

### PR DESCRIPTION
**Pull request by offthewall123**

### CODE REVIEW

File: cpp/src/plasma/vmemcache_store.cc [hunk 1]
```diff
@@ -28,13 +28,46 @@
...
+using namespace tinyxml2;
+
 #define CACHE_MAX_SIZE (1024 * 1024 * 1024L)
 #define CACHE_EXTENT_SIZE 512
 
 namespace plasma {
 
+boolean detectInitailPath() {
+  XMLDocument doc;
+  XMLError loaderr = doc.LoadFile("persistent-memory.xml");
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:42]: Do we still need this configuration ```persistent-memory.xml```? Since we already have path passing, we should be able to detect the remaining size for those dirs and check whether we can allocate that much. @jikunshang

File: cpp/src/plasma/vmemcache_store.cc [hunk 2]
```diff
@@ -28,13 +28,46 @@
...
+    numanode = numanode->NextSiblingElement();
+  }
+  return true;
+}
+
 // Connect here is like something initial
 Status VmemcacheStore::Connect(const std::string& endpoint) {
+  if(!detectInitailPath()) {
+      ARROW_LOG(FATAL)<<"Initial vmemcache failed!";
+      return Status::UnknownError("Initial vmemcache failed!");
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:69]: We should provide the detailed information why Vmemcache initialization failed. In this case, the err msg should be "Fail to provide the enough size for allocation" or "Directories not exist" and you can get related information the return code from method ```detectInitailPath```.

File: cpp/src/plasma/vmemcache_store.cc (hunk 1 above)
Comment (jikunshang) [reply cpp/src/plasma/vmemcache_store.cc:42]: I think a configuration file is needed, the reason is same with OAP. Currently, pmem dir is hardcoded, and it will not be same in different machines. Besides, we can use a disk dir as external store.

File: cpp/src/plasma/vmemcache_store.cc [hunk 3]
```diff
@@ -28,29 +28,91 @@
...
 // Connect here is like something initial
 Status VmemcacheStore::Connect(const std::string& endpoint) {
-  auto size_start = endpoint.find("size:") + 5;
-  auto size_end = endpoint.size();
-  std::string sizeStr = endpoint.substr(size_start, size_end);
-  unsigned long long size = std::stoull(sizeStr);
-  if (size == 0) size = CACHE_MAX_SIZE;
-  totalCacheSize = size * totalNumaNodes;
-  ARROW_LOG(DEBUG) << "vmemcache size is " << size;
+  std::vector<numaNodeInfo> numaNodeVt;
```
Comment (jikunshang) [inline cpp/src/plasma/vmemcache_store.cc:99]: We can rename it as `numaNodeInfos`.

File: cpp/src/plasma/vmemcache_store.cc [hunk 4]
```diff
@@ -28,30 +28,80 @@
...
+
+  totalNumaNodes = numaNodeVt.size();
   for (int i = 0; i < totalNumaNodes; i++) {
     // initial vmemcache on numa node i
+    numaNodeInfo nninfo = numaNodeVt[i];
     VMEMcache* cache = vmemcache_new();
     if (!cache) {
       ARROW_LOG(FATAL) << "Initial vmemcache failed!";
-      return Status::UnknownError("Initial vmemcache failed!");
+      return Status::UnknownError("Initial vmemcache failed! 2222");
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:99]: Remove 2222 comment please.

File: cpp/src/plasma/vmemcache_store.cc [hunk 5]
```diff
@@ -28,30 +28,80 @@
...
+
+#include "plasma/tools/tinyxml2.h"
+
 #define CACHE_MAX_SIZE (1024 * 1024 * 1024L)
 #define CACHE_EXTENT_SIZE 512
 
 namespace plasma {
+bool VmemcacheStore::detectInitailPath(std::vector<numaNodeInfo> &numaNodeVt) {
+  tinyxml2::XMLDocument doc;
+  tinyxml2::XMLError loaderr = doc.LoadFile("/tmp/persistent-memory.xml");
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:43]: @jikunshang could we pass in some configurations instead of xml file or direct in starting command?

File: cpp/src/plasma/vmemcache_store.h [hunk 6]
```diff
@@ -93,13 +103,15 @@ class VmemcacheStore : public ExternalStore {
...
+  bool detectInitailPath(std::vector<numaNodeInfo> &numaNodeVt);
+
  private:
   void Evict(std::vector<ObjectID>& ids, std::vector<std::shared_ptr<Buffer>>& datas);
   std::vector<VMEMcache*> caches;
   std::vector<std::shared_ptr<numaThreadPool>> putThreadPools;
   std::vector<std::shared_ptr<numaThreadPool>> getThreadPools;
   int totalNumaNodes = 2;
-  int threadInPools = 12;
+  // int threadInPools = 12;
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.h:114]: Why we need to comment thsi out?

**Comment by winningsix** [comment]
A quick question: do we need to copy tinyxml2.cpp file into our source? @jikunshang what's the practice for Arrow?

File: cpp/src/plasma/vmemcache_store.cc (hunk 5 above)
Comment (jikunshang) [reply cpp/src/plasma/vmemcache_store.cc:43]: have discussed with @offthewall123 , there are some ways: 
1. hard code config(current impl) 
2. xml config file, 
3. pass some parameters via command, 
4. pass nothing and using os API to detect Pmem path and size. 
Solution 3 is not fancy since it may pass a lot things and not easy to parse a string.
We will choose to combine solution 2 and 4.

**Comment by jikunshang** [comment]
> A quick question: do we need to copy tinyxml2.cpp file into our source? @jikunshang what's the practice for Arrow?

I am not sure about Arrow practice, but I think it's fine to add some third party source code, it's what Redis ae did. Otherwise, we should add a third party library as a cmake module.
Besides, this xml lib are using zlib license, have some risk or conflict?

File: cpp/src/plasma/vmemcache_store.cc [hunk 7]
```diff
@@ -30,54 +30,48 @@
...
-    tinyxml2::XMLElement* path = numanode->FirstChildElement();   
-    tinyxml2::XMLElement* requiredSize = path->NextSiblingElement();
-    tinyxml2::XMLElement* readPoolSize = requiredSize->NextSiblingElement();   
-    tinyxml2::XMLElement* writePoolSize = readPoolSize->NextSiblingElement();
-
-    numaNodeInfo info;
+  
+  int numanodeNum = configMap.size()/5;
+  for(int i = 1; i <= numanodeNum; i++){
+    uint64_t requiredSize = std::stoull(configMap["requiredSize" +  std::to_string(i)]);
```
Comment (jikunshang) [inline cpp/src/plasma/vmemcache_store.cc:51]: Can we abstract a method like ConvertConfigMapToNumaNodeInfo ?

File: cpp/src/plasma/tools/getProperties.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
+#include "getProperties.h"
```
Comment (jikunshang) [inline cpp/src/plasma/tools/getProperties.cpp:1]: Add Apache license.

File: cpp/src/plasma/tools/getProperties.cpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+#include <fstream>
+#include <iostream>
+ 
+using namespace std;
+ 
+bool isSpace(char c)
+{
+	if (' ' == c || '\t' == c)
```
Comment (jikunshang) [inline cpp/src/plasma/tools/getProperties.cpp:9]: Arrow cpp code style use 2 spaces as indent.

File: cpp/src/plasma/tools/getProperties.cpp [hunk 10]
```diff
@@ -0,0 +1,90 @@
+#include "getProperties.h"
+#include <fstream>
+#include <iostream>
+ 
+using namespace std;
+ 
+bool isSpace(char c)
```
Comment (jikunshang) [inline cpp/src/plasma/tools/getProperties.cpp:7]: Abstract a parse class?

File: cpp/src/plasma/CMakeLists.txt [hunk 11]
```diff
@@ -133,7 +133,7 @@ if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
   set_property(SOURCE dlmalloc.cc APPEND_STRING PROPERTY COMPILE_FLAGS " -Wno-conversion")
 endif()
 
-list(APPEND PLASMA_EXTERNAL_STORE_SOURCES "external_store.cc" "hash_table_store.cc" "vmemcache_store.cc")
+list(APPEND PLASMA_EXTERNAL_STORE_SOURCES "external_store.cc" "hash_table_store.cc" "vmemcache_store.cc" "tools/getProperties.cpp")
```
Comment (jikunshang) [inline cpp/src/plasma/CMakeLists.txt:136]: We can remove this cpp file

Comment (winningsix) [inline cpp/src/plasma/CMakeLists.txt:136]: rename it to Plasma.proporties.

File: cpp/src/plasma/tools/getProperties.h [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+#define COMMENT_CHAR '#'
+#include <fstream>
+#include <map>
+#include <string>
+#include <vector>
+#include "../plasma/vmemcache_store.h"
+
+using namespace std;
+
+class getProperties {
```
Comment (winningsix) [inline cpp/src/plasma/tools/getProperties.h:28]: Minor: could we use any existing library instead implementing our own? https://stackoverflow.com/questions/874052/properties-file-library-for-c-or-c

File: cpp/src/plasma/tools/getProperties.h [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+  int start_pos = 0, end_pos = line.size() - 1, pos;
+  if ((pos = line.find(COMMENT_CHAR)) != -1) {
+    if (0 == pos) {
+      return false;
+    }
+    end_pos = pos - 1;
+  }
+  string new_line = line.substr(start_pos, start_pos + 1 - end_pos);
+
+  if ((pos = new_line.find('=')) == -1) return false;
```
Comment (winningsix) [inline cpp/src/plasma/tools/getProperties.h:121]: Since we define the way of parsing, we should provide some configuration template somewhere in our source code.

File: cpp/src/plasma/vmemcache_store.cc [hunk 14]
```diff
@@ -28,30 +28,66 @@
...
+#include <map>
+
+#include "plasma/tools/getProperties.h"
+
 #define CACHE_MAX_SIZE (1024 * 1024 * 1024L)
 #define CACHE_EXTENT_SIZE 512
 
 namespace plasma {
+bool VmemcacheStore::DetectInitailPath(std::vector<numaNodeInfo>& numaNodeInfos) {
+  string path = "/tmp/persistent-memory.properties";
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:42]: Any way to avoiding hardcore path?

File: cpp/src/plasma/vmemcache_store.cc [hunk 15]
```diff
@@ -28,30 +28,66 @@
...
+    if (retVal < 0) {
+      ARROW_LOG(FATAL) << "Directory: " << info.initialPath << " not exist.";
+      return false;
+    }
+    uint64_t blockSize = pathInfo.f_bsize;
+    uint64_t freeSize = pathInfo.f_bfree * blockSize;
+    if (info.requiredSize > freeSize) {
+      ARROW_LOG(WARNING) << "Failed to provide enough size for allocation";
+      ARROW_LOG(WARNING) << "Directory: " << info.initialPath
+                         << "will use max freesize: " << freeSize << "B";
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:63]: space between word free and size

File: cpp/src/plasma/vmemcache_store.cc [hunk 16]
```diff
@@ -28,30 +28,66 @@
...
       ARROW_LOG(FATAL) << "Initial vmemcache failed!";
       return Status::UnknownError("Initial vmemcache failed!");
     }
-    // TODO: how to find path and bind numa?
-    std::string s = "/mnt/pmem" + std::to_string(i);
-    ARROW_LOG(DEBUG) << "initial vmemcache on " << s << ", size" << size
+    std::string path = nninfo.initialPath;
+    u_int64_t size = nninfo.requiredSize;
+
+    ARROW_LOG(DEBUG) << "initial vmemcache on " << path << ", size" << size
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:90]: Let's log numa node info as well.

File: cpp/src/plasma/tools/getProperties.h [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+  static bool isSpace(char c);
+};
+
+std::vector<numaNodeInfo> getProperties::ConvertConfigMapToNumaNodeInfo(
+    map<string, string>& configMap) {
+  std::vector<numaNodeInfo> res;
+  int numanodeNum = configMap.size() / 5;
+  for (int i = 1; i <= numanodeNum; i++) {
+    numaNodeInfo info;
+    info.requriedSize = std::stoull(configMap["requiredSize" + std::to_string(i)]);
```
Comment (winningsix) [inline cpp/src/plasma/tools/getProperties.h:47]: We should consider the case only passing in paths with absent of all other values empty. In another word, we should think of the default value when possible.

File: cpp/src/plasma/tools/getProperties.h [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+#include <fstream>
+#include <map>
+#include <string>
+#include <vector>
+#include "../plasma/vmemcache_store.h"
+#include "arrow/util/logging.h"
+
+using namespace std;
+
+class getProperties {
```
Comment (jikunshang) [inline cpp/src/plasma/tools/getProperties.h:32]: It's better to use upper case to define a class name

File: cpp/src/plasma/tools/getProperties.h [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+#include <mntent.h>
+#include <sys/vfs.h>
+#include <fstream>
+#include <map>
+#include <string>
+#include <vector>
+#include "../plasma/vmemcache_store.h"
+#include "arrow/util/logging.h"
+
+using namespace std;
```
Comment (jikunshang) [inline cpp/src/plasma/tools/getProperties.h:30]: remove this line, and add namespace plasma, since this class is for plasma module internal use.

File: cpp/src/plasma/tools/getProperties.h [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+bool getProperties::getDefaultConfig(std::vector<plasma::numaNodeInfo>& numanodeInfos) {
+  FILE* mount_table = NULL;
+  struct mntent* mount_entry;
+  mount_table = setmntent("/etc/mtab", "r");
+  if (!mount_table) {
+    ARROW_LOG(FATAL) << "set mount entry error";
+    return false;
+  }
+  int numaNodeId = 0;
+  while (1) {
```
Comment (jikunshang) [inline cpp/src/plasma/tools/getProperties.h:55]: while (true)

File: cpp/src/plasma/vmemcache_store.cc [hunk 21]
```diff
@@ -260,7 +305,7 @@ Status VmemcacheStore::Get(const std::vector<ObjectID>& ids,
   return Status::OK();
 }
 
-// not used
+// not usedtotalCacheSize
```
Comment (jikunshang) [inline cpp/src/plasma/vmemcache_store.cc:308]: revert this change please

File: cpp/src/plasma/vmemcache_store.h [hunk 22]
```diff
@@ -93,13 +103,14 @@ class VmemcacheStore : public ExternalStore {
...
   void Metrics(int64_t* memory_total, int64_t* memory_used) override;
 
+  bool DetectInitailPath(std::vector<numaNodeInfo> &numaNodeVt, std::string path);
+
  private:
   void Evict(std::vector<ObjectID>& ids, std::vector<std::shared_ptr<Buffer>>& datas);
   std::vector<VMEMcache*> caches;
   std::vector<std::shared_ptr<numaThreadPool>> putThreadPools;
   std::vector<std::shared_ptr<numaThreadPool>> getThreadPools;
   int totalNumaNodes = 2;
```
Comment (jikunshang) [inline cpp/src/plasma/vmemcache_store.h:113]: default value should not be 2.

File: cpp/src/plasma/sampleConf/persistent-memory-sample.properties [hunk 23]
```diff
@@ -0,0 +1,11 @@
+numanodeId0 = 0
```
Comment (jikunshang) [inline cpp/src/plasma/sampleConf/persistent-memory-sample.properties:1]: rename it to `persistent-memory.properties.template`

File: cpp/src/plasma/tools/getProperties.h (hunk 12 above)
Comment (jikunshang) [reply cpp/src/plasma/tools/getProperties.h:32]: import some third party lib may bring license conflict, right?

File: cpp/src/plasma/tools/getProperties.h [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+    std::string mount_point = mount_entry->mnt_dir;
+    struct statfs pathInfo;
+    if (mount_point.find("/mnt/pmem") != -1) {
+      if (statfs(mount_point.c_str(), &pathInfo) < 0) {
+        return false;
+      }
+      plasma::numaNodeInfo info;
+      info.numaNodeId = numaNodeId;
+      info.initialPath = mount_point;
+      info.readPoolSize = 12;
```
Comment (jikunshang) [inline cpp/src/plasma/tools/getProperties.h:72]: avoid magic number, and some machines may not have many cores, it's better to calculate a value like `getTotalCores / numanodeNum / 2`  .

File: cpp/src/plasma/sampleConf/persistent-memory-sample.properties (hunk 23 above)
Comment (winningsix) [inline cpp/src/plasma/sampleConf/persistent-memory-sample.properties:1]: Let's rename the dir from ```sampleConf``` to ```conf``` and rename the file from ```persistent-memory-sample.properties``` to ```persistent-memory-sample.properties.template```.

File: cpp/src/plasma/sampleConf/persistent-memory-sample.properties [hunk 25]
```diff
@@ -0,0 +1,11 @@
...
+initialPath0 = /mnt/pmem0
+requiredSize0 = 1500000
+readPoolSize0 = 12
```
Comment (winningsix) [inline cpp/src/plasma/sampleConf/persistent-memory-sample.properties:4]: Since we're using it as template, not need to have it as concrete value. You can refer to what Spark did.
https://github.com/apache/spark/blob/master/conf/spark-defaults.conf.template

File: cpp/src/plasma/tools/getProperties.h [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+bool GetProperties::getDefaultConfig(std::vector<plasma::numaNodeInfo>& numanodeInfos) {
+  FILE* mount_table = NULL;
+  struct mntent* mount_entry;
+  mount_table = setmntent("/etc/mtab", "r");
+  if (!mount_table) {
+    ARROW_LOG(FATAL) << "set mount entry error";
+    return false;
+  }
+  int numaNodeId = 0;
+  while (true) {
```
Comment (winningsix) [inline cpp/src/plasma/tools/getProperties.h:56]: Why do we need while true here? Can we use for each?

File: cpp/src/plasma/conf/persistent-memory.properties.template [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+# numanodeId0 = 0
+# initialPath0 = /mnt/pmem0
+# requiredSize0 = 1500000
+# readPoolSize0 = 12 
+# writePoolSize0 = 12
+
+# numanodeId1 = 1
+# initialPath1 = /mnt/pmem1
+# requiredSize1 = 1500000
+# readPoolSize1 = 12
```
Comment (winningsix) [inline cpp/src/plasma/conf/persistent-memory.properties.template:31]: @jikunshang  could we use a shared pool here for different numa node? We can logically change the size for each thread pool when handling requests from client.

File: cpp/src/plasma/vmemcache_store.cc [hunk 28]
```diff
@@ -28,31 +28,80 @@
...
-  if (size == 0) size = CACHE_MAX_SIZE;
-  totalCacheSize = size * totalNumaNodes;
-  ARROW_LOG(DEBUG) << "vmemcache size is " << size;
+  std::vector<numaNodeInfo> numaNodeInfos;
+  std::string configPath = "";
+  int configPathStart = endpoint.find("configPath");
+  if (configPathStart != -1) {
+    configPath = endpoint.substr(configPathStart + 11, endpoint.size());
+  }
+  if (!DetectInitailPath(numaNodeInfos, configPath)) {
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:82]: typo: initial

File: cpp/src/plasma/tools/PlasmaProperties.h [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+#include <map>
+#include <string>
+#include <vector>
+#include "../plasma/vmemcache_store.h"
+
+namespace plasma {
+
+class PlasmaProperties {
+ public:
+  static bool parseConfig(const std::string& filename,
```
Comment (winningsix) [inline cpp/src/plasma/tools/PlasmaProperties.h:33]: Let's directly use numaNodeInfos instead of configMap. Then you can remove the ```convertConfigMapToNumaNodeInfo``` method.

File: cpp/src/plasma/vmemcache_store.cc [hunk 30]
```diff
@@ -28,31 +28,80 @@
...
+      ARROW_LOG(FATAL) << "Directory: " << info.initialPath << " not exist.";
+      return false;
+    }
+    uint64_t blockSize = pathInfo.f_bsize;
+    uint64_t freeSize = pathInfo.f_bfree * blockSize;
+    if (info.requiredSize > freeSize) {
+      ARROW_LOG(WARNING) << "Failed to provide enough size for allocation";
+      ARROW_LOG(WARNING) << "Directory: " << info.initialPath
+                         << "will use max free_size * " << DEFALUT_CONFIG_MULTIPLIER
+                         << " " << freeSize << "B";
```
Comment (winningsix) [inline cpp/src/plasma/vmemcache_store.cc:67]: Can we use MB instead?

Comment (offthewall123) [reply cpp/src/plasma/vmemcache_store.cc:67]: But the  persistent-memory.properties takes B as the unit.

File: cpp/src/plasma/vmemcache_store.cc [hunk 31]
```diff
@@ -28,31 +28,73 @@
...
 // Connect here is like something initial
 Status VmemcacheStore::Connect(const std::string& endpoint) {
-  auto size_start = endpoint.find("size:") + 5;
-  auto size_end = endpoint.size();
-  std::string sizeStr = endpoint.substr(size_start, size_end);
-  unsigned long long size = std::stoull(sizeStr);
-  if (size == 0) size = CACHE_MAX_SIZE;
-  totalCacheSize = size * totalNumaNodes;
-  ARROW_LOG(DEBUG) << "vmemcache size is " << size;
+  std::string argStr = endpoint.substr(12, endpoint.length());
```
Comment (jikunshang) [inline cpp/src/plasma/vmemcache_store.cc:76]: avoid magic number, define a const int  instead.

File: cpp/src/plasma/tools/PlasmaProperties.cpp [hunk 32]
```diff
@@ -0,0 +1,192 @@
...
+
+bool PlasmaProperties::isCommentChar(char c) {
+  if (c == COMMENT_CHAR) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+void PlasmaProperties::trim(std::string& str) {
```
Comment (winningsix) [inline cpp/src/plasma/tools/PlasmaProperties.cpp:132]: Could we use ```boost::algorithm::trim(s);``` instead?
Or you can do like 
```#include <iostream>
#include <string>
#include <regex>

std::string ltrim(const std::string& s) {
	return std::regex_replace(s, std::regex("^\\s+"), std::string(""));
}

std::string rtrim(const std::string& s) {
	return std::regex_replace(s, std::regex("\\s+$"), std::string(""));
}

std::string trim(const std::string& s) {
	return ltrim(rtrim(s));
}```

FYI https://www.techiedelight.com/trim-string-cpp-remove-leading-trailing-spaces/

File: cpp/src/plasma/tools/PlasmaProperties.cpp [hunk 33]
```diff
@@ -0,0 +1,192 @@
...
+    info.writePoolSize =
+        std::stoul(this->getProperty("writePoolSize" + std::to_string(i)));
+    info.requiredSize =
+        std::stoull(this->getProperty("requiredSize" + std::to_string(i)));
+    this->numanodeInfos.push_back(info);
+  }
+  return totalNumaNodeNum == (int)numanodeInfos.size();
+}
+
+std::string PlasmaProperties::getDefaultProperty(std::string key) {
```
Comment (winningsix) [inline cpp/src/plasma/tools/PlasmaProperties.cpp:105]: Let's define those default values in one place. Otherwise, it's not easy to maintain the code.

File: cpp/src/plasma/tools/PlasmaProperties.cpp [hunk 34]
```diff
@@ -0,0 +1,192 @@
...
+#include "unistd.h"
+
+namespace plasma {
+
+PlasmaProperties::PlasmaProperties(std::string& argStr, std::string& propertyFilePath_) {
+  this->argsStr = argStr;
+  this->propertyFilePath = propertyFilePath;
+}
+
+void PlasmaProperties::parseArgStr(std::string argStr) {
```
Comment (winningsix) [inline cpp/src/plasma/tools/PlasmaProperties.cpp:35]: Let's have some tests on this.

Comment (offthewall123) [reply cpp/src/plasma/tools/PlasmaProperties.cpp:36]: These are private methods.
I only make one public method here.

File: cpp/src/plasma/CMakeLists.txt [hunk 35]
```diff
@@ -254,9 +255,10 @@ if(ARROW_BUILD_SHARED)
 else()
   set(PLASMA_TEST_LIBS plasma_static ${PLASMA_STATIC_LINK_LIBS})
 endif()
-
+     
+add_plasma_test(test/plasma_properties_tests EXTRA_LINK_LIBS ${PLASMA_TEST_LIBS})
 add_plasma_test(test/serialization_tests EXTRA_LINK_LIBS ${PLASMA_TEST_LIBS})
-add_plasma_test(test/client_tests
+add_plasma_test(test/client_tests
```
Comment (jikunshang) [inline cpp/src/plasma/CMakeLists.txt:261]: revert this.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 36]
```diff
@@ -266,6 +266,7 @@ else()
     BOOST_SOURCE_URL
     # These are trimmed boost bundles we maintain.
     # See cpp/build_support/trim-boost.sh
+    "https://github.com/jikunshang/boost/releases/download/boost-1.67.0/boost_1_67_0.tar.gz"
```
Comment (winningsix) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:269]: Could we remove this or at least using OAP asset link?

File: cpp/src/plasma/test/plasma_properties_tests.cc [hunk 37]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+TEST_F(TestPlasmaProperties, ParseArgStrTest) {
+  std::string propertyStr = "";
+  std::string argStr =
+      "totalNumaNodeNum:2,numaNodeId1:1,initialPath1:/mnt/"
+      "pmem0,requiredSize1:15000000,readPoolSize1:12,writePoolSize1:12?numaNodeId2:2,"
+      "initialPath2:/mnt/pmem1,requiredSize2:1500000,readPoolSize2:12,writePoolSize2:12";
+  PlasmaProperties* p = new PlasmaProperties(argStr, propertyStr);
+  p->getNumaNodeInfos();
```
Comment (winningsix) [inline cpp/src/plasma/test/plasma_properties_tests.cc:50]: Should we check the result to see whether it's under the expectation?

**Comment by offthewall123** [comment]
jira link
https://jira.devtools.intel.com/browse/POAE7-298

**Comment by winningsix** [comment]
LGTM +1

**Comment by winningsix** [comment]
cc @jikunshang

**Comment by jikunshang** [comment]
LGTM
